File: 239-Sliding Window Maximum-滑动窗口最大值/239-Sliding Window Maximum-滑动窗口最大值.java
class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
    	int len = nums.length;
    	int[] res = new int[len - k + 1];
    	MonotonousQueue monotonousQueue = new MonotonousQueue();
    	for(int i = 0; i < k; i++) {
    		monotonousQueue.add(nums[i]);
    	}
    	res[0] = monotonousQueue.peek();
    	if(res.length == 1) return res;
    	for(int i = k; i < nums.length; i++) {
            monotonousQueue.poll(nums[i - k]);
    		monotonousQueue.add(nums[i]);
    		res[i - k + 1] = monotonousQueue.peek();
    	}
    	return res;
    }
}

public class MonotonousQueue{
	LinkedList<Integer> queue;
	
	public MonotonousQueue() {
		this.queue = new LinkedList<>();
	}



	public void add(int value) {
		while(!queue.isEmpty() && queue.getLast() < value) {
			queue.removeLast();
		}
		queue.add(value);
	}
	
	public void poll(int value) {
		if(!queue.isEmpty() && queue.peek() == value) {
			queue.poll();
		}
	}
	
	public int peek() {
		return queue.peek();
	}
}

File: 239-Sliding Window Maximum-滑动窗口最大值/239-Sliding Window Maximum-滑动窗口最大值.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
 var maxSlidingWindow = function(nums, k) {
    let res = []
    let monotonousQueue = []
    if(nums.length === 1) return nums

    for(let i = 0; i < k; i++){
        while(monotonousQueue.length !== 0 && nums[i] > monotonousQueue[monotonousQueue.length - 1]){
            monotonousQueue.pop()
        }
        monotonousQueue.push(nums[i])
    }

    res.push(monotonousQueue[0])

    for(let i = k; i < nums.length; i++){
        // 出队列
        if(monotonousQueue.length !== 0 && nums[i -  k] === monotonousQueue[0]){
            monotonousQueue.shift()
        }

        // 入队列
        while(monotonousQueue.length !== 0 && nums[i] > monotonousQueue[monotonousQueue.length - 1]){
            monotonousQueue.pop()
        }
        monotonousQueue.push(nums[i])

        res.push(monotonousQueue[0])
    }
    return res
};